nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/process/ProcessUtils.java

```diff
@@ -230,7 +230,11 @@ public static boolean waitFor(Supplier<Boolean> sign, Duration timeout, boolean
                         System.out.flush();
                     }
                 }
-                Thread.yield();
+                try {
+                    Thread.sleep(100L);
+                } catch (InterruptedException e) {
+                    LOG.log(TRACE, "Interrupted while waiting", e);
```

Comment:
Do not swallow `InterruptedException`.

Reply:
Also 100 ms of sleeping is quite a lot. It will slow down integration tests - I believe Thread.yield is enough. Those ports should be closed immediately, if they are not, there is a bug somewhere. Is it visible on Linux too?

Reply:
Yet one idea - something in your system (firewall, antivirus) might be attaching to every open port and then they are not closed immediately when Java closes them. I have no idea how to resolve it reliably except your workaround (sleeping between iterations). Maybe we should make the sleep value configurable (system property).

But maybe we should start with some unit test - creating a fake server, running 1 000 000 iterations in parallel thread and then closing the server. We have a windows machine in GitHub Actions, linux on Jenkins, it should be reproducible if it is a common problem. I can try that in few days, tomorrow I have to work on other things.

But I really don't like solutions like "take a pause for this magical number of millis and all issues vanish".

Reply:
I confirm that on Ubuntu 24.04.1 LTS at least.

Reply:
> Do not swallow `InterruptedException`.

I've modified the code not to swallow `InterruptedException`.

Reply:
> Those ports should be closed immediately, if they are not, there is a bug somewhere.

The behavior that TCP ports remains in state TIME_WAIT for a while (i.e. not immediately closed) comes from the TCP specification. See RFC 9293, Section 3.3.2 [1].

In Linux systems, TCP ports in TIME_WAIT state is kept for 60 seconds by default [2].
In Windows systems, 120 seconds by default [3].

Although it is possible to immediately close the TCP ports (without transitioning to TIME_WAIT state) by using the [SO_LINGER](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/net/SocketOptions.html#SO_LINGER) option, but it is not recommended because it forcibly close the connection by RST and may violate the TCP quiet time restriction.

[1] https://www.rfc-editor.org/rfc/rfc9293.html#section-3.3.2
[2] https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt
[3] https://learn.microsoft.com/en-us/biztalk/technical-guides/settings-that-can-be-modified-to-improve-network-performance

Reply:
Aha, `SO_LINGER` rings the bell to me. However in our case we don't use the TCP for communication, we just check that server is listening; another option could be to wait until server closes the successfully established connection.

> The TCP state transition diagrams above both show orderly connection termination. There’s another way to terminate a TCP connection and that’s by aborting the connection and sending an RST rather than a FIN. This is usually achieved by setting the SO_LINGER socket option to 0. This causes pending data to be discarded and the connection to be aborted with an RST rather than for the pending data to be transmitted and the connection closed cleanly with a FIN. It’s important to realise that when a connection is aborted any data that might be in flow between the peers is discarded and the RST is delivered straight away; usually as an error which represents the fact that the “connection has been reset by the peer”. The remote peer knows that the connection was aborted and neither peer enters TIME_WAIT.

...

> it’s inefficient to rapidly open and close TCP connections to the same server so it makes sense beyond the issue of TIME_WAIT to try and maintain connections for longer periods of time rather than shorter periods of time. 

https://serverframework.com/asynchronousevents/2011/01/time-wait-and-its-design-implications-for-protocols-and-scalable-servers.html